### PR TITLE
[Relay] Prepare DeadCodeElimination for running post LowerTEPass/ManifestAlloc.

### PR DESCRIPTION
**Pull request by mbs-octoml**
As part of #9483 we need to prepare some critical Relay passes for running after
lowering and conversion to DPS. For DCE we need to make sure we never remove
side-effecting let-bound expressions, such as for allocation or evaluation of
an external function with unknown effectfulness.

Introduce a new purity pre-pass. It makes a half-hearted attempt at accounting
for functions by tracking both 'eval' and 'call' purity, but must fallback to
assuming call-impurity in more difficult cases (eg calling a function passed as
a parameter, calling a function projected from a tuple, etc). However it seems
plenty good enough.

Purity must also be accounted for when determining the usage count of let-bound
variables, so reworked that. Collapsed the let-bound value accumulation pass into
the usage counting pass to make up for inserting the new purity analysis pass.

Note that some unit tests assume DCE eliminates dead writes to references. The
previous implementation certainly did that, but also eliminated all writes. To do this
soundly will require a bit of liveness analysis. Rather than doing that now I've added
an 'ignore_impurity' flag to DeadCodeElimination() to recover the (unsound) legacy
behavior.

### CODE REVIEW

File: src/relay/op/memory/memory.h [hunk 1]
```diff
@@ -35,6 +35,7 @@ namespace tvm {
 namespace relay {
 
 Expr AllocStorage(Expr size, Expr alignment, SEScope se_scope, DataType dtype_hint);
+const Op& MemoryAllocTensorOp();
```
Comment (jroesch) [inline src/relay/op/memory/memory.h:38]: This looks kind of odd, can you provide a comment at least?

**Review by jroesch — APPROVED** [review]
It would be good to get a second set of eyes but looks good to me. Modulo my comment.

**Comment by mikepapadim** [comment]
LGTM

File: src/relay/op/memory/memory.h (hunk 1 above)
Comment (mbs-octoml) [reply src/relay/op/memory/memory.h:38]: Oh it snuck in from the main change, sorry. Added comment, just trying to cut down on the duplicated special operator name strings.

**Review by electriclilies — APPROVED** [review]
Really nice and readable! Caught a few spelling errors though :)

File: include/tvm/relay/transform.h [hunk 2]
```diff
@@ -540,7 +550,7 @@ TVM_DLL Function ToCPS(const Function& f, const IRModule& mod);
 /*!
  * \brief Remove the continuation argument of a CPS function.
  *
- * Note that this only transform the type back into un-CPS form
+ * Note that this only transform the type back into un-CPS formA
```
[the remark this answers is not in the document]
Comment (mbs-octoml) [reply include/tvm/relay/transform.h]: thanks, done!

File: src/relay/transforms/dead_code.cc [hunk 3]
```diff
@@ -18,158 +18,565 @@
...
+  Purity VisitExpr_(const ConstantNode*) final { return {/*pure_eval=*/true, /*pure_call=*/true}; }
+
+  Purity VisitExpr_(const ConstructorNode*) final {
+    return {/*pure_eval=*/true, /*pure_call=*/true};
+  }
+
+  Purity VisitExpr_(const OpNode* op_node) final {
+    // Primitive operators are pure unless marked as 'stateful'.
+    static OpAttrMap<bool> attr_map = Op::GetAttrMap<TOpIsStateful>("TOpIsStateful");
+    bool is_statefull = attr_map.count(GetRef<Op>(op_node)) && attr_map[GetRef<Op>(op_node)];
```
[the remark this answers is not in the document]
Comment (mbs-octoml) [reply src/relay/transforms/dead_code.cc]: done

File: src/relay/transforms/dead_code.cc [hunk 4]
```diff
@@ -18,158 +18,565 @@
...
+
+  Purity VisitExpr_(const RefCreateNode*) final {
+    // The creation of the  ref itself is unobservable other than via the reads/writes into it.
+    return {/*pure_eval=*/true, /*pure_call=*/true};
+  }
+
+  Purity VisitExpr_(const RefWriteNode* ref_write_node) final {
+    Purity ref_purity = VisitExpr(ref_write_node->ref);
+    ICHECK(ref_purity.pure_call);  // reference is first-order
+    // The call purity of the written value is lost [RULE D].
```
[the remark this answers is not in the document]
Comment (mbs-octoml) [reply src/relay/transforms/dead_code.cc:265]: added a comment to explain -- gotta accumulate any let-bindings inside the value.

**Comment by tmoreau89** [comment]
Thank you @mbs-octoml @mikepapadim @electriclilies @jroesch ; the PR has been merged

**Pull request merged**